core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala

```diff
@@ -312,7 +313,51 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
     timeout.max(member.rebalanceTimeoutMs)
   }
 
-  def generateMemberIdSuffix = UUID.randomUUID().toString
+  def generateMemberId(clientId: String,
+                       groupInstanceId: Option[String]): String = {
+    groupInstanceId match {
+      case None =>
+        clientId + MEMBER_ID_DELIMITER + UUID.randomUUID().toString
+      case Some(instanceId) =>
+        instanceId + MEMBER_ID_DELIMITER + currentStateTimestamp.get
+    }
+  }
+
+  // This validation doesn't handle the case where a static member changes
+  // its group.instance.id while maintains the same member.id. That edge case should
+  // be caused by a hacked consumer client since current client logic doesn't
+  // support group.instance.id change on runtime.
+  def validateMemberIdIfStatic(memberId: String): Boolean = {
+    val delimiterIdx = memberId.lastIndexOf(MEMBER_ID_DELIMITER)
+    // Member id must have formatted with valid delimiter as [prefix]-[suffix] to
+    // be validated.
+    if (delimiterIdx == -1) {
+      info(s"given member.id $memberId was ill-formatted without delimiter '-'.")
+      true
+    } else {
+      val memberIdPrefix = Some(memberId.substring(0, delimiterIdx))
+      val isKnownStaticMember = hasStaticMember(memberIdPrefix)
```

Comment:
What if the client.id happens to be the same as some of the existing group.instance.id? I'd prefer a stricter check to also check the second part (whether it is a UUID or timestamp).

Reply:
That's a valid concern. Let me check how to do that.